Custom entitlements: add README and other improvements (#1167)
RevenueCat/purchases-android: examples/CustomEntitlementComputationSample/README.md

The SDK should only be configured once the initial appUserID is known.

## Using the app

To use the app, you should do the following: 
- Configure your app in the [RevenueCat dashboard](https://app.revenuecat.com/). No special configuration is needed, but you should contact RevenueCat support
before enabling this mode to ensure that it's the right one for your app. It's highly recommended to set Transfer Behavior to "Keep with original App User ID" in the RevenueCat Dashboard. 
- Update the API key in Constants.kt. You can update the default `appUserID` there too, although apps in this mode should 
always be calling configure only when the appUserID is already known. 
- Update the applicationId in defaultConfig of the app-level build.gradle to match your RevenueCat app configuration.
- Have at least one Offering with at least one Package configured for Android, since this is the one that the purchase button will use. 

Once configured correctly, the app will allow you to log in with different users, and will show a list of all the times CustomerInfoListener fired, as well as 
the values for each one. 

Happy testing!

> ![sample screenshot](./Sample%20screenshot.png)

Author: just realized that this screenshot has my name, I'll update on Monday

Author: we can use this one
<img width="422" alt="Screenshot 2023-07-24 at 2 34 30 PM" src="https://github.com/RevenueCat/purchases-android/assets/3922667/8c21f79b-73e2-4c56-9935-30057f28304c">

Reviewer: Took another one with more padding in the `CustomerInfo listener values` section and included it

## Using Custom Entitlements mode

### Installation: 

This package is available on Maven and can be included via Gradle.

To use this mode, ensure that you install the customEntitlementsComputation build flavor by specifying the dependency as:
```gradle

> revenuecat = { module = "com.revenuecat.purchases:purchases@customEntitlementComputation", version.ref = "purchases" }

Author: could use some android-expert eyes on this

Reviewer 2: This will have a different name once it's published. Also, this nomenclature is only if they use gradle catalogs... I would kinda vote to use simple gradle `implementation "com.revenuecat.purchases..."`

```
for Gradle version catalogs, or
```gradle

> implementation 'com.revenuecat.purchases:purchases@customEntitlementsComputation:6.8.0'

Reviewer: `@customEntitlementsComputation` would be valid if we were publishing as a flavor, but we are publishing the flavor as a different artifact

```
implementation 'com.revenuecat.purchases:purchases-custom-entitlement-computation:6.8.0'
```

I would also remove the Gradle catalogs version and simply leave the standard nomenclature

```kotlin
val builder = PurchasesConfiguration.Builder(
    this,
    Constants.GOOGLE_API_KEY,
    Constants.defaultAppUserID
)

Purchases.configure(builder.build())
```

### Getting Offerings: 

Call getOfferings through either the Async / Await or completion blocks alternatives:

```kotlin

val offerings = Purchases.sharedInstance.awaitOfferings()

```

### Switching users: 

To switch to a different user, call:

```kotlin
Purchases.sharedInstance.switchUser(newUserID)
```

This will ensure that all purchases made from this point on are posted for the new appUserID. 
After calling this method, you might need to call your backend to refresh entitlements for the new appUserID if they haven't been refreshed already.

### Making purchases:

Call `awaitPurchase()`:

```kotlin
try {
    val (transaction, customerInfo) =
        Purchases.sharedInstance.awaitPurchase(purchaseParams)
    // refresh entitlements with your backend
} catch (error: PurchasesTransactionException) {
    if (error.userCancelled) {
        _uiState.update { it.copy(displayErrorMessage = "User cancelled") }
    } else {
        _uiState.update { it.copy(displayErrorMessage = error.message) }

>     }

Author: we should add the exact code to catch: 
- receipt already in use
- payment pending
- purchase cancelled
- product already purchased

Reviewer 2: Hmm they could check the `PurchasesTransactionException.error.errorCode` and check for the following error codes:
- `PurchasesErrorCode.ProductAlreadyPurchasedError`
- `PurchasesErrorCode.PurchaseCancelledError`: though it's easier to just use the `userCancelled` property.
- `PurchasesErrorCode.ReceiptAlreadyInUseError`: I might be missing something but this one shouldn't happen I think?  As in, we are posting purchases:
  -  Performed through the app (which should only be for new purchases and so should never have this error)
  -  On app foreground, if we haven't synced the purchase before. In this case, it would fail silently to sync, so a user in this mode would never see this happening.
- `PurchasesErrorCode.PaymentPendingError`: need to test to confirm this error happens...

```kotlin
            val purchaseParams = PurchaseParams.Builder(activity, aPackage).build()
try {
    val (transaction, customerInfo) =
        Purchases.sharedInstance.awaitPurchase(purchaseParams)
    // refresh entitlements with your backend
} catch (error: PurchasesTransactionException) {
    if (error.userCancelled) {
        _uiState.update { it.copy(displayErrorMessage = "Purchase was cancelled by the user") }
    } else {
        val errorMessage = when (error.code) {
            PurchasesErrorCode.ReceiptAlreadyInUseError ->
                "The receipt is already in use by another subscriber. " +
                        "Log in with the previous account or contact support " +
                        "to get your purchases transferred to regain access"
            PurchasesErrorCode.PaymentPendingError ->
                "The purchase is pending and may be completed at a later time. " +
                        "This can happen when awaiting parental approval or going " +
                        "through extra authentication flows for credit cards " +
                        "in some countries"
            PurchasesErrorCode.ProductAlreadyPurchasedError ->

>                 "This product is already active for the user."

Author: sounds good.

            PurchasesErrorCode.PurchaseNotAllowedError ->
                "Purchasing wasn't allowed, which is common if the card is declined " +
                        "or the purchase is not available in the country " +
                        "you're trying to purchase from."
            PurchasesErrorCode.StoreProblemError ->
                "There was a problem with the Google Play Store. This is a generic " +
                        "Google error, and there's not enough information to " +
                        "determine the cause."
            else -> "FAILED TO PURCHASE: ${error.message}"
        }
        _uiState.update { it.copy(displayErrorMessage = errorMessage) }

    }
}
```

### Observing changes to purchases:

To ensure that your app reacts to changes to subscriptions in real time, you can use `customerInfoStream`. This stream will only fire when new `customerInfo` is registered

> in RevenueCat, like when a subscription is renewed. If there are no changes from the last value, it will not fire. This means it's not guaranteed to fire on every app open.

Author: yeah, good call, we should update

> in RevenueCat, like when a subscription is renewed. If there are no changes from the last value, it will not fire. This means it's not guaranteed to fire on every app open.

Reviewer: also, `customerInfoStream` is not a thing in Android. I will update

> in RevenueCat, like when a subscription is renewed. If there are no changes from the last value, it will not fire. This means it's not guaranteed to fire on every app open.

Reviewer: check it out and let me know what you think

# Custom Entitlements Computation Example App

This app is useful for testing RevenueCat under Custom Entitlements Computation mode and understanding how it works.

## What is Custom Entitlements Computation mode? 

> This is a special behavior mode for RevenueCat SDK, is intended for apps that will do their own entitlement computation separate from RevenueCat. 

Author: ```suggestion
This is a special behavior mode for RevenueCat SDK. It's intended for apps that will do their own entitlement computation separate from RevenueCat. 
```

Apps using this mode rely on webhooks to signal their backends to refresh entitlements with RevenueCat.

> In this mode, RevenueCat will not generate anonymous user IDs, it will not refresh customerInfo cache automatically only when a purchase goes through 

Author: ```suggestion
In this mode, RevenueCat will not generate anonymous user IDs, or refresh customerInfo cache automatically. CustomerInfo is only refreshed when a purchase goes through 
```

and it will disallow all methods other than those for configuration, switching users, getting offerings and making purchases.

> When in this mode, the app should use switchUser() to switch to a different App User ID if needed. 

Author: ```suggestion
When in this mode, the app should use `switchUser()` to switch to a different App User ID if needed. 
```

The SDK should only be configured once the initial appUserID is known.

## Using the app

To use the app, you should do the following: 
- Configure your app in the [RevenueCat dashboard](https://app.revenuecat.com/). No special configuration is needed, but you should contact RevenueCat support
before enabling this mode to ensure that it's the right one for your app. It's highly recommended to set Transfer Behavior to "Keep with original App User ID" in the RevenueCat Dashboard. 
- Update the API key in Constants.kt. You can update the default `appUserID` there too, although apps in this mode should 
always be calling configure only when the appUserID is already known. 
- Update the applicationId in defaultConfig of the app-level build.gradle to match your RevenueCat app configuration.
- Have at least one Offering with at least one Package configured for Android, since this is the one that the purchase button will use. 

> Once configured correctly, the app will allow you to log in with different users, and will show a list of all the times CustomerInfoListener fired, as well as 

Author: ```suggestion
Once configured correctly, the app will allow you to log in with different users, and will show a list of all the times `CustomerInfoListener` fired, as well as 
```
